qCC/ccCommandLineParser.cpp

```diff
@@ -390,6 +390,200 @@ QString ccCommandLineParser::exportEntity(	CLEntityDesc& entityDesc,
 	return (result != CC_FERR_NO_ERROR ? QString("Failed to save result in file '%1'").arg(outputFilename) : QString());
 }
 
+bool ccCommandLineParser::selectClouds(bool selectAll, bool reverse, bool selectFirst, bool selectLast, bool selectRegex, int firstNr, int lastNr, QRegExp regex)
+{
+	if (selectRegex && !regex.isValid())
+	{
+		return error(QObject::tr("Regex string invalid: %1").arg(regex.errorString()));
+	}
+
+	//store everthyng in the rest vector
+	while (!m_clouds.empty())
+	{
+		m_clouds_rest.push_back(m_clouds.front());
+		m_clouds.erase(m_clouds.begin());
+	}
+
+	//sort the rest clouds by uniqueID (it will sort the clouds as it were loaded/created)
+	std::sort(m_clouds_rest.begin(), m_clouds_rest.end(), [](CLCloudDesc a, CLCloudDesc b) { return (a.pc->getUniqueID() < b.pc->getUniqueID()); });
+
+	//put elements to the front facing vector
+	int index = 0;
+	size_t lastIndex = m_clouds_rest.size() - 1;
+	for (std::vector<CLCloudDesc>::iterator it = m_clouds_rest.begin(); it != m_clouds_rest.end();)
+	{
+		QString nameToValidate = QObject::tr("%1/%2").arg(it->basename).arg(it->pc->getName());
+		bool required = false;
+		if (!reverse)
+		{
+			//first {n}
+			if (selectFirst && index < firstNr)
+			{
+				required = true;
+			}
+
+			//last {n}
+			if (selectLast && index > lastIndex - lastNr)
+			{
+				required = true;
+			}
+		}
+		else
+		{
+			//not first {n}
+			if (selectFirst && index >= firstNr && !selectLast)
+			{
+				required = true;
+			}
+
+			//not last {n}
+			if (selectLast && index <= lastIndex - lastNr && !selectFirst)
+			{
+				required = true;
+			}
+
+			//not first and not last
+			if (selectFirst && selectLast && index >= firstNr && index <= lastIndex - lastNr)
+			{
+				required = true;
+			}
+		}
+
+		//regex has higher priority than first/last overwrite
+		if (selectRegex)
+		{
+			if (regex.indexIn(nameToValidate) > -1)
+			{
+				//regex matched
+				required = !reverse;
+			}
+			else
+			{
+				//regex not matched
+				required = reverse;
+			}
+		}
+
+		//selectAll has higher priority than first/last/regex overwrite
+		if (selectAll)
+		{
+			required = !reverse;
+		}
+
+		if (required)
+		{
+			print(QObject::tr("\t[*] UID: %2 name: %1").arg(nameToValidate).arg(it->pc->getUniqueID()));
+			m_clouds.push_back(*it);
+			it = m_clouds_rest.erase(it);
+		}
+		else
+		{
+			print(QObject::tr("\t[ ] UID: %2 name: %1").arg(nameToValidate).arg(it->pc->getUniqueID()));
+			++it;
+		}
+		index++;
+	}
+	return true;
+}
+
+bool ccCommandLineParser::selectMeshes(bool selectAll, bool reverse, bool selectFirst, bool selectLast, bool selectRegex, int firstNr, int lastNr, QRegExp regex)
```

Comment:
Could you merge the code of the 2 methods in one with a local templated function and passing the 2 vectors m_clouds/m_unselectedClouds or m_meshes/m_unselectedMeshes as parameters?

I can do it if it's not clear.

Reply:
Well i'll push fixes, and then i' feel like this is over my knowledge.

I've tried to merge them but the main issue i had is how to replace the m_clouds with m_meshes. I've tried to use a pointer, but without succes. I'm not really sure if i know everything about pointers. If it is easy for you to direct me in that direction than i could try it, but otherwise. I'm not gonna able to do that.

Reply:
Ok, it's 'easy' with a templated function. I'll do it when you'll have updated the branch.

Reply:
I'll take a look at templated functions. If i'd try that can i define anywhere or should i place it to a special place like at the top or bottom or a template location in the class?

Reply:
The simplest would be to declare it as a static method, before your 2 member methods (so that each can call the templated one).

It will be something like:
template<class EntityDesc > static bool SelectEntities(bool selectAll, bool reverse, bool selectFirst, bool selectLast, bool selectRegex, int firstNr, int lastNr, QRegExp regex, std::vector<EntityDesc>& selectedEntities, std::vector<EntityDesc>& unselectedEntities)
{
}

Reply:
should i define a param struct instead a horde of params? similar to globalShiftParams?
I've started it but i reverted, because it created loads of unnecessary typing.

Reply:
As you prefer! Since no other method will call this function, I think we can leave it like that, but a struct is always cleaner.